cpp/src/arrow/dataset/filter.cc

```diff
@@ -758,43 +797,107 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
+                                                        const Schema& schema) {
+  struct {
+    struct Validated {
+      std::shared_ptr<Expression> expr;
+      std::shared_ptr<DataType> type;
+    };
+
+    Result<Validated> Validate(const Expression& expr) {
+      Validated out;
+      ARROW_ASSIGN_OR_RAISE(out.expr, InsertImplicitCasts(expr, schema_));
+      ARROW_ASSIGN_OR_RAISE(out.type, out.expr->Validate(schema_));
+      return std::move(out);
+    }
 
-  void operator()(const NotExpression& expr) { VisitExpression(*expr.operand(), *this); }
+    Result<std::shared_ptr<Expression>> operator()(const NotExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto op, Validate(*expr.operand()));
 
-  void operator()(const AndExpression& expr) {
-    VisitExpression(*expr.left_operand(), *this);
-    VisitExpression(*expr.right_operand(), *this);
-  }
+      if (op.type->id() != Type::BOOL) {
+        op.expr = op.expr->CastTo(boolean()).Copy();
+      }
+      return not_(std::move(op.expr));
+    }
 
-  void operator()(const OrExpression& expr) {
-    VisitExpression(*expr.left_operand(), *this);
-    VisitExpression(*expr.right_operand(), *this);
-  }
+    Result<std::shared_ptr<Expression>> operator()(const AndExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto lhs, Validate(*expr.left_operand()));
+      ARROW_ASSIGN_OR_RAISE(auto rhs, Validate(*expr.right_operand()));
 
-  void operator()(const ComparisonExpression& expr) {
-    VisitExpression(*expr.left_operand(), *this);
-    VisitExpression(*expr.right_operand(), *this);
-  }
+      if (lhs.type->id() != Type::BOOL) {
+        lhs.expr = lhs.expr->CastTo(boolean()).Copy();
+      }
+      if (rhs.type->id() != Type::BOOL) {
+        rhs.expr = rhs.expr->CastTo(boolean()).Copy();
+      }
+      return and_(std::move(lhs.expr), std::move(rhs.expr));
+    }
 
-  template <typename E>
-  void operator()(const E& expr) const {}
+    Result<std::shared_ptr<Expression>> operator()(const OrExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto lhs, Validate(*expr.left_operand()));
+      ARROW_ASSIGN_OR_RAISE(auto rhs, Validate(*expr.right_operand()));
 
-  std::vector<std::string> fields;
-};
+      if (lhs.type->id() != Type::BOOL) {
+        lhs.expr = lhs.expr->CastTo(boolean()).Copy();
+      }
+      if (rhs.type->id() != Type::BOOL) {
+        rhs.expr = rhs.expr->CastTo(boolean()).Copy();
+      }
+      return or_(std::move(lhs.expr), std::move(rhs.expr));
+    }
+
+    Result<std::shared_ptr<Expression>> operator()(const ComparisonExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto lhs, Validate(*expr.left_operand()));
+      ARROW_ASSIGN_OR_RAISE(auto rhs, Validate(*expr.right_operand()));
+
+      if (lhs.type->Equals(rhs.type)) {
+        return expr.Copy();
+      }
+
+      if (lhs.expr->type() == ExpressionType::SCALAR) {
+        lhs.expr = lhs.expr->CastTo(rhs.type).Copy();
+      } else {
+        rhs.expr = rhs.expr->CastTo(lhs.type).Copy();
```

Comment:
Should you check that `rhs.expr` is a scalar here?

Reply:
Even if rhs.expr is not a scalar, it must be cast to lhs.type to enable comparison

Reply:
Ok. But in that case is there a specific reason to cast rhs to lhs' type rather than lhs to rhs' type?

(I expect that ideally, this would depend on the actual types, e.g. if lhs is `int32` and rhs is `int64`, you want to cast lhs to rhs' type)

Reply:
There isn't a specific reason; fully general promotion is just harder and I planned to leave it for a follow up PR.